chain/src/txhashset/bitmap_accumulator.rs

```diff
@@ -0,0 +1,241 @@
+		let mut chunk = BitmapChunk::new();
+
+		let mut leaves = idx.into_iter().peekable();
+		while let Some(x) = leaves.peek() {
+			if *x < chunk_idx * 1024 {
+				// skip until we reach our first chunk
+				leaves.next();
+			} else if *x < (chunk_idx + 1) * 1024 {
+				let idx = leaves.next().expect("next after peek");
+				chunk.set(idx % 1024, true);
+			} else {
+				self.append_chunk(chunk)?;
+				chunk_idx += 1;
+				chunk = BitmapChunk::new();
+			}
+		}
+		if chunk.any() {
+			self.append_chunk(chunk)?;
+		}
+		debug!(
+			"applied {} chunks from idx {} to idx {} ({}ms)",
+			1 + chunk_idx - from_chunk_idx,
+			from_chunk_idx,
+			chunk_idx,
+			now.elapsed().as_millis(),
+		);
+		Ok(())
+	}
+
+	/// Apply updates to the bitmap accumulator given an iterator of invalidated idx and
+	/// an additional iterator of new idx to be added.
+	/// We determine the existing chunks to be rebuilt given the invalidated idx.
+	/// We then add new idx, extending the accumulator with new chunk(s) as necessary.
+	pub fn apply<T, U>(&mut self, invalidated_idx: T, new_idx: U) -> Result<(), Error>
+	where
+		T: IntoIterator<Item = u64>,
+		U: IntoIterator<Item = u64>,
+	{
+		// Determine the earliest chunk by looking at the min invalidated idx (assume sorted).
+		// Rewind prior to this and reapply new_idx.
+		// Note: We rebuild everything after rewind point but much of the bitmap may be
+		// unchanged. This can be further optimized by only rebuilding necessary chunks and
+		// rehashing.
+		if let Some(first_chunk_idx) = invalidated_idx.into_iter().next().map(|x| x / 1024) {
+			self.rewind_prior(first_chunk_idx)?;
+			self.pad_left(first_chunk_idx)?;
+			self.apply_from(new_idx, first_chunk_idx)?;
+		}
+
+		// Trim any "empty" chunks from the right to keep everything consistent.
+		self.trim_right()?;
+
+		Ok(())
+	}
+
+	/// Given the provided chunk idx rewind the bitmap accumulator to the end of the
+	/// previous chunk ready for the updated chunk to be appended.
+	fn rewind_prior(&mut self, chunk_idx: u64) -> Result<(), Error> {
+		let last_pos = self.backend.size();
+		let mut pmmr = PMMR::at(&mut self.backend, last_pos);
+		let chunk_pos = pmmr::insertion_to_pmmr_index(chunk_idx + 1);
+		let rewind_pos = chunk_pos.saturating_sub(1);
+		pmmr.rewind(rewind_pos, &Bitmap::create())
+			.map_err(|e| ErrorKind::Other(e))?;
+		Ok(())
+	}
+
+	/// Make sure we append empty chunks to fill in any gap before we append the chunk
+	/// we actually care about. This effectively pads the bitmap with 1024 chunks of 0s
+	/// as necessary to put the new chunk at the correct place.
+	fn pad_left(&mut self, chunk_idx: u64) -> Result<(), Error> {
+		let last_pos = self.backend.size();
+		let chunk_pos = pmmr::insertion_to_pmmr_index(chunk_idx);
+		for _ in last_pos..chunk_pos {
+			self.append_chunk(BitmapChunk::new())?;
+		}
+		Ok(())
+	}
+
+	/// Recursively trim the last chunk if empty. Stop when rightmost chunk is non-empty
+	/// or the accumulator itself is empty.
```

Comment:
What does "empty" mean? Is that all 0s? I.e. all spent outputs? But for every spent output there's a later unspent output, so how does this trimming work?

Reply:
Yes exactly - empty here means all zero values.

I actually went back and forth between these two competing (and somewhat ambiguous) interpretations -
a) bitmap is the txo size rounded up to multiple of 1024
b) bitmap is minimum required to cover the final 1 value, rounded up to multiple of 1024

I ended up going with (b) for a couple of reasons - 
1) roaring bitmap actually only represents 1 values, so without explicitly providing the txo size its undefined exactly how big the bitmap needs to be
2) in practice this question is moot - the txo set always includes a final 1 value because the situation where we create an output and then spend it in the _same_ block would not result in a 0 value due to block level cut-through. So the last pos in the bitmap is always a 1.

So keeping the bitmap minimum size is easier to implement and does not require the txo size to be explicitly specified when building bitmaps.

Reply:
but does 0 mean spent or unspent? if 0 is spent then a) and b) come out the same, since the very last txo is always unspent.

Reply:
1 is unspent, 0 spent

Reply:
If the last pos is always a 1 (unspent), then why is there a need to trim at all?

Reply:
I think I did this so we did not have undefined behavior around the bitmap itself. We only have a "last pos is always 1" due to our block specific behavior. There is nothing intrinsic to the bitmap accumulator that would ensure the last pos will always be a 1.

Its possible to write test cases where this "last pos is always 1" does not hold and its nice to have defined unambiguous behavior here.

Reply:
Rethinking this. Lets make the `size` simply the index of the rightmost 1 value in the bitmap. 

So we rely on the assumption that our usage involves an unspent output at the rightmost position of the output MMR. But this should simplify how we go about making sure the bitmap is in a canonical state with no "empty" chunks incorrectly appended at the end.

Our usage guarantees this - a new block always contains at least one unspent output (the coinbase output). Any other outputs that appear after this in canonical output ordering must always be unspent also, otherwise they would have been cut-through.

And we can reflect this in the various tests by simply making sure the rightmost bit is always set to true.

Reply:
> the index of the rightmost 1 value in the bitmap.

Isn't that 1 less than the number of txos? If so, I would just use that instead, as it's an actual size.
No need for special casing pre-genesis either, when there is no rightmost 1 value.

Reply:
When increasing this value, we can simply do an assert than the prior bit is 1...